javascript/computer_science/project_hash_map.md

```diff
@@ -0,0 +1,227 @@
+
+Remember we don't want collisions, in a perfect world each bucket will either have 0 or 1 Node only, so we grow our buckets to have more chance that our Nodes will spread and not stack up in the same buckets. To grow our buckets, we create a new buckets list that is double the size of the old buckets list, then we retrieve all nodes from the old buckets and insert them into the new buckets.
+
+#### When do we know that it's time to grow our buckets size
+
+To deal with this, our hash map class need to keep track of two new fields, the `capacity` and the `load factor`.
+
+- The `capacity` is the amount of buckets we currently have. Keeping track of this will let us know if our map has reached the threshold, the threshold will depend on our factor variable as well.
+
+- The `load factor` is a number that we can assign our hash map to at the start. It's the factor that will determine when is it a good time to grow our buckets, for example a load factor of `0.75` means our hash map will need to grow its buckets when the capacity reaches 75% full. Setting it too low will consume too much memory by having too many empty buckets, while setting it too high will allow our buckets to have collisions before we grow them. Usually a good balance of `0.75 to 1` is used.
+
+### Computation complexity
+
+A hash map is very efficient in its insertion, retrieving and removing operations. That is because we use array indexes to do those operations. A hash map has an average case complexity of `O(1)` for the following methods:
+- Insertion
+- Retrieving
+- Deletion
+
+Assuming we have a good hash map written. The worst case of those operations would be `O(n)` and that happens when we have all our data hashes to the same exact bucket.
+
+The Growth of our hash map has the complexity of `O(n)` at all times.
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+
+#### Self restriction and limitation
+
+  Before we get started, we need to lay down some ground rules. As we all know JavaScript is a dynamic language the restriction we have in other languages does not exist in JavaScript for example if you create an array of size `16` to be your buckets size, nothing stopping you from storing items at index `500`. This defeats the purpose we are trying to demonstrate, so we need to put some self restriction to work around this.
+
+  1. When you create a function that takes a hash code value and returns the bucket index, use this snippet before the function returns to make sure it's accessing inbound array indexes only.
+
+      ```javascript
+      if (index < 0 || index >= buckets.length) {
+        throw new Error("Trying to access index out of bound");
+      }
+      ```
+
+  1. Inside the class DO NOT use Objects to store your key value pairs, And DO NOT use the array method `push` to insert your nodes into the buckets, always use indexes.
```

Comment:
> Also, is it okay to store the key/value pair as an object on the linked list nodes?

Since Node is not part of the hash map and using an object literal is not violating the rule, but I still feel that this takes the magic away. In the end of the day, using this will still make us give our key values to JS built in library to handle for us. That's what we're trying to avoid, and building our own key value mechanics.